applications/FluidDynamicsApplication/custom_elements/data_containers/qs_vms/qs_vms_derivative_utilities.cpp

```diff
@@ -0,0 +1,413 @@
+
+    KRATOS_DEBUG_ERROR_IF(rdNdX.size1() != number_of_nodes)
+        << "rdNdX not initialized properly. [ required rdNdX.size1() = " << number_of_nodes
+        << ", rdNdX.size1() = " << rdNdX.size1() << " ].\n";
+
+    KRATOS_DEBUG_ERROR_IF(rNodalVelocity.size2() != 2)
+        << "rNodalVelocity.size2() != TDim [ TDim = " << 2
+        << ", rNodalVelocity.size2() = " << rNodalVelocity.size2() << " ].\n";
+
+    KRATOS_DEBUG_ERROR_IF(rdNdX.size2() != 2)
+        << "rdNdX.size2() != TDim [ TDim = " << 2
+        << ", rdNdX.size2() = " << rdNdX.size2() << " ].\n";
+
+    noalias(rOutput) = ZeroVector(3);
+    for (IndexType i = 0; i < number_of_nodes; ++i) {
+        rOutput[0] += rdNdX(i, 0) * rNodalVelocity(i, 0);
+        rOutput[1] += rdNdX(i, 1) * rNodalVelocity(i, 1);
+        rOutput[2] += rdNdX(i, 0) * rNodalVelocity(i, 1) + rdNdX(i, 1) * rNodalVelocity(i, 0);
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <>
+void QSVMSDerivativeUtilities<3>::CalculateStrainRate(
+    Vector& rOutput,
+    const Matrix& rNodalVelocity,
+    const Matrix& rdNdX)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(rOutput.size() != 6)
+        << "rOutput is not properly initialized. [ rOutput.size() != 6, "
+           "rOutput.size() = "
+        << rOutput.size() << " ].\n";
+
+    const IndexType number_of_nodes = rNodalVelocity.size1();
+
+    KRATOS_DEBUG_ERROR_IF(rdNdX.size1() != number_of_nodes)
+        << "rdNdX not initialized properly. [ required rdNdX.size1() = " << number_of_nodes
+        << ", rdNdX.size1() = " << rdNdX.size1() << " ].\n";
+
+    KRATOS_DEBUG_ERROR_IF(rNodalVelocity.size2() != 3)
+        << "rNodalVelocity.size2() != TDim [ TDim = " << 3
+        << ", rNodalVelocity.size2() = " << rNodalVelocity.size2() << " ].\n";
+
+    KRATOS_DEBUG_ERROR_IF(rdNdX.size2() != 3)
+        << "rdNdX.size2() != TDim [ TDim = " << 3
+        << ", rdNdX.size2() = " << rdNdX.size2() << " ].\n";
+
+    noalias(rOutput) = ZeroVector(6);
+    for (IndexType i = 0; i < number_of_nodes; ++i) {
+        rOutput[0] += rdNdX(i,0) * rNodalVelocity(i,0);
+        rOutput[1] += rdNdX(i,1) * rNodalVelocity(i,1);
+        rOutput[2] += rdNdX(i,2) * rNodalVelocity(i,2);
+        rOutput[3] += rdNdX(i,0) * rNodalVelocity(i,1) + rdNdX(i,1) * rNodalVelocity(i,0);
+        rOutput[4] += rdNdX(i,1) * rNodalVelocity(i,2) + rdNdX(i,2) * rNodalVelocity(i,1);
+        rOutput[5] += rdNdX(i,0) * rNodalVelocity(i,2) + rdNdX(i,2) * rNodalVelocity(i,0);
+    }
+
+    KRATOS_CATCH("");
+}
+
+template <>
+void QSVMSDerivativeUtilities<2>::CalculateStrainRateVelocityDerivative(
+    Vector& rOutput,
+    const IndexType DerivativeNodeIndex,
+    const IndexType DerivativeDirectionIndex,
+    const Matrix& rdNdX)
+{
+    KRATOS_TRY
+
+    KRATOS_DEBUG_ERROR_IF(rOutput.size() != 3)
+        << "rOutput is not properly initialized. [ rOutput.size() != 3, "
+           "rOutput.size() = "
+        << rOutput.size() << " ].\n";
```

Comment:
Same comment. I think this is not require. You allocate either outside the function or reallocate below when assigning to a `ZeroVector(3)`.

Reply:
same comment as above